Exception stacks

c42f opened the pull request:
# An exception stack system for julia

Here's my attempt to fix #12485 / #19979 by introducing an exception stack
system. The basic problem is that you may throw a new exception while handling
an exception.  What do we do then? The existing runtime simply overwrites the
first exception, loosing any information it might have had about the root
cause.

A distinct problem discussed in #12485 is that task switching looses all
exception and backtrace information. I've tackled this at the same time
though in hindsight it might have been better to do the exception stack
first, followed by fixing task switching.

## Design

To solve the first problem, I've implemented an exception stack mechanism.
Conceptually:

* Every `throw()` pushes an `(exception,backtrace)` pair onto a stack of current
  exceptions.
* When you exit a `catch` block normally, any exception generated within the
  corresponding `try` is considered to have been handled. The stack is therefore
  popped back to the state at the entry of the corresponding `try`.
* The full stack is available for root cause analysis via `Base.catch_stack()`.

For example:

```julia
try
    error("A")
catch
    try
        error("B")
    catch
        for (e, bt) in Base.catch_stack()
            showerror(stdout, e, bt)
            println()
        end
    end
end
```

Produces

```
A
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] top-level scope at ./REPL[3]:2
 [3] eval(::Module, ::Any) at ./boot.jl:319
 [4] eval_user_input(::Any, ::REPL.REPLBackend) at /home/chris/dev/julia/usr/share/julia/stdlib/v1.0/REPL/src/REPL.jl:85
 [5] macro expansion at /home/chris/dev/julia/usr/share/julia/stdlib/v1.0/REPL/src/REPL.jl:117 [inlined]
 [6] (::getfield(REPL, Symbol("##28#29")){REPL.REPLBackend})() at ./task.jl:259
B
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] top-level scope at ./REPL[3]:5
 [3] eval(::Module, ::Any) at ./boot.jl:319
 [4] eval_user_input(::Any, ::REPL.REPLBackend) at /home/chris/dev/julia/usr/share/julia/stdlib/v1.0/REPL/src/REPL.jl:85
 [5] macro expansion at /home/chris/dev/julia/usr/share/julia/stdlib/v1.0/REPL/src/REPL.jl:117 [inlined]
 [6] (::getfield(REPL, Symbol("##28#29")){REPL.REPLBackend})() at ./task.jl:259
```

## Implementation

#### Lowering

A new lowered expression head `:pop_exc` is introduced and emitted in any
location where a catch block exits normally (either by falling out the bottom,
or using return, break, goto). This pops the exception stack back to the state
of the associated `enter`. There needs to be some way to track the associated
`enter` in the IR, so I chose to emit an SSAValue token from `enter` which is
consumed by `pop_exc`. I tried various alternatives, but this was by far the
nicest in terms of non-disruptive integration into the SSAIR processing code,
and supporting both the interpreter and codegen.

`finally` lowering no longer includes a `foreigncall`, so expressions using
finally can now be interpreted. This includes the output of `@test` so may
give a nice speed boost.

#### Runtime

* The exception in transit is no longer a single pointer in the thread local
  storage. It's stored in `ptls->current_task->exc_stack` instead, along with
  associated backtrace. Space for the exception and backtrace is allocated
  on demand inside `throw_internal`, after changing GC mode.

* The exception stack is ~~copied out onto the `Task`~~ naturally preserved during task switching.
  I've been back and forth on several variations here, as the exact detail
  of how this is done depends on how the memory for the exception
  stack is allocated (GC vs plain malloc).

* `jl_current_exception()` is now the method to retreive the current exception
  from within a `JL_CATCH` dynamic context.

* `_resetstkoflw()` handling is consolidated into `jl_eh_restore_state`.
  Needs proper testing on windows.

* Several places where manual restoration of the exception and backtrace was
  done are no longer necessary because the stack system handles these
  automatically.

* An exception stack type, `jl_exc_stack_t` and associated manipulation
  functions has been added. Conceptually stores a stack of
  (exception,backtrace) pairs. It's stored in a shared buffer so we can avoid
  reallocating when throwing and catching the same exception or set of
  exceptions repeatedly.

* `jl_eh_restore_state` has become non-inline. It seemed good to get this out
  of the public header.

* `jl_sig_throw` is now used with `jl_throw_in_ctx` from signal handlers, to
  make the special circumstances clear and to avoid conflation with rethrow
  which now simply rethrows the existing exception stack.

#### Interpreter / Codegen

Some minor rearrangement in the interpreter to make the eh buffer management
and `leave` handling a little easier to understand as I found this surprisingly
confusing.

Mostly small changes in codegen. Representing the enter token as an SSAValue
really integrates well here when you need to retreive it for the associated
`pop_exc`.


#### GC

What? How did this rabbit hole lead down into the GC?

Unfortunately the raw backtrace buffers contain references to GC allocated
objects. Since these buffers can now live on a `Task`, their contents need to
be GC rooted. Unfortunately the raw buffers are also in a non-julia native
format, so I had to write specialized GC code for them.

I'm not sure this was a good idea, especially having extra code to maintain in
this rather sensitive part of the codebase. Perhaps it would be better to
modify our raw backtrace buffer format so that it can be understood by the GC
without major changes.


## TODO

* [x] Check some potential GC issues I found in testing.
* [x] Decide on detail of `Base.catch_stack()` API and export it.
* [x] Double check consolidated `_resetstkoflw` handling on windows (checked on cross compiled version via wine on ubuntu)
* [x] Tests for exception preservation during task switching
* [x] Reinstate rooting of temporary `bt_data`
* [x] Check that access to backtraces for failed `Task`s still works correctly and figure out how to access the full exception stack for those.

#### Optional

* [x] ~~Reconsider best practices for `rethrow(exc)` vs just throwing a new exception~~. Can be done separately
* [x] `jl_apply_with_saved_exception_state` may be unnecessary now
* [x] `jl_rethrow_other` ~~is probably unnecessary now~~ is required for `rethrow(ex)`
* [x] Persist backtrace when task exits with an exception

c42f committed:
Some cleanup of indentiation and comments

c42f committed:
Lowering for exception stacks

* A new lowered expression head :pop_exc is introduced and emitted in any
  location where a catch block exits normally (either by stepping out,
  or using return, break, goto). The semantics of :pop_exc are to pop
  the exception stack back to the state of the associated enter.

* Make Expr(:enter) return a token which may be consumed by :pop_exc,
  thereby allowing the interpreter and codegen to know which :enter
  state should be used to pop the exception stack.  I tried various
  alternatives for this association, but this was by far the nicest in
  terms of non-disruptive integration into the SSAIR processing code,
  and supporting both the interpreter and codegen.

c42f committed:
Add exception stack system to runtime

Runtime
-------

* An exception stack type, `jl_exc_stack_t` and associated manipulation functions
  has been added. Conceptually this stores a stack of (exception,backtrace)
  pairs. It's stored in a contiguous buffer so we can avoid reallocating when
  throwing and catching the same exception or set of exceptions repeatedly. Space
  for the exception and backtrace is allocated on demand inside `throw_internal`,
  after changing GC mode. Several variations were tried for allocating this
  sgorage, including allocating up front with malloc on the thread local state
  and copying during task switching. Keeping everything on the task seemed the
  simplest as it involves the least copying and keeps track of the exception
  stack buffers in a unified way.

* The exception in transit is no longer a single pointer in the thread local
  storage. It's stored in `ptls->current_task->exc_stack` instead, along with
  associated backtrace.

* `jl_current_exception()` is now the method to retreive the current exception
  from within a `JL_CATCH` dynamic context.

* Several places where manual restoration of the exception and backtrace was
  done are no longer necessary because the stack system handles these
  automatically. This code is removed, including
  `jl_apply_with_saved_exception_state`.

* `jl_eh_restore_state` has become non-inline. It seemed good to get this out
  of the public header.

* `jl_sig_throw` is now used with `jl_throw_in_ctx` from signal handlers, to
  make the special circumstances clear and to avoid conflation with rethrow
  which now simply rethrows the existing exception stack.

* Make rethrow outside a catch block into an error.

* Use `ptls->previous_exception` to support `jl_exception_occurred` in
  embedding API.

* finally lowering no longer includes a `foreigncall`, so expressions
  using finally can now be interpreted.

Interpreter / Codegen
---------------------

Mostly small changes here, to track the `:enter` and `:pop_exc` association
with the SSAValue token. The token SSAValue slot is ideal here for storing the
state of the exception stack at the `:enter`.

GC
--

Integrate exception and raw backtrace scanning as a special case into GC mark
loop. This is necessary now that Task objects can refer to exception and
backtrace data in raw form.

c42f committed:
catch_stack() for julia level exception stack access + tests

iamed2 commented:
I won't comment on the technical aspects of the solution but I like it conceptually! And I'm happy to see those issues resolved, they made error handling in Dispatcher and stacktrace generation in Memento tricky business.

c42f commented:
@iamed2 Hah yes I originally came across this when working on logging.

To add to the commentary above, One of my biggest difficulties with this patch was the need to GC root objects in the bt buffer, and also GC the buffer itself in the case that it's attached to a `Task`. Combine this with the need to actually fill the buffer from inside a signal handler (and potentially other restricted contexts) and you're in a tricky situation. There's probably several holes I've yet to plug there.

Actually I hope to delete most of the GC changes which were done here as they seem like overkill. My current best idea is to use the new efficient array-of-small-`Union` support to represent the raw backtrace buffer as a julia array of `Union{Ptr{Nothing},UInt,CodeInfo,...}` which could be filled without allocation and with only a little more book keeping than what we currently do. Need to sort out whether we can call `jl_gc_wb` in the contexts where we fill the buffer.

Is this a sound idea? Advice on how to improve the memory management in this patch would be particularly welcome.

c42f commented:
Ok, CI failures are because I've got something wrong in GC land and the debug build is failing on an assertion. My best guess is that I've broken whatever invariants `jl_gc_alloc_buf` are meant to obey.

Anyway, the release build does "work" so the branch is in a state where the curious can play with it until the real problem is fixed.

c42f commented:
Looking into alternative storage for backtrace buffers, the `Union` isn't going to work because we really need to store pointers in there. So perhaps we're stuck with two parallel arrays instead like the current `bt`, `bt2` which appears in some places. Or just put up with having some special purpose code in the gc.

c42f committed:
Consolidate _resetstkoflw handling

The logic for this was repeated in three quite different ways, in
JL_CATCH, interpreter and codegen but putting it into the runtime in
jl_eh_restore_state should be sufficient.

Also move jl_eh_restore_state out of main header - seems unnecessary to
expose such implementation detail, and inconsistent with other eh
functions.

c42f commented on src/gc.c:
```diff
@@ -1901,7 +1901,7 @@ excstack: {
         while (itr > 0) {
             size_t bt_size = jl_exc_stack_bt_size(exc_stack, itr);
-            uint64_t *bt_data = jl_exc_stack_bt_data(exc_stack, itr);
+            uintptr_t *bt_data = jl_exc_stack_bt_data(exc_stack, itr);
```
Hah, this was a hilariously stupid bug which cost a lot of time due to randomly busted GC on 32 bit, I don't know what I was thinking using `uint64_t` here :-/

It would have been caught early if we were able to use `-Werror` on CI builds but I suspect that is quite disruptive in other ways.

StefanKarpinski replied:
> It would have been caught early if we were able to use -Werror on CI builds but I suspect that is quite disruptive in other ways.

Please, we should absolutely do this. We've had a creeping increase of compiler warnings in the past month or so.

c42f commented:
Hmm, just encountered a spurious test failure on 32 bit due to https://github.com/JuliaLang/julia/issues/27788.

c42f commented:
Tests have finally passed at least once on all systems and I've done a fair amount of cleanup. Should be good for a closer look.

I've exported `catch_stack()` for now, it could do with a better name / better cohesion with the `catch_backtrace()` API.

c42f commented:
Ok, tests have passed after restarting the appveyor x86_64 build three times. There is an intermittent failure in a single one of the spawn tests on that platform related to deleting a directory, see https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.30766/job/wnxk1pxa27xte6un for example.

This seems to be the test from https://github.com/JuliaLang/julia/issues/26687 coming back again after https://github.com/JuliaLang/julia/pull/28107 was merged.

The fact that it's recurring here more frequently could be related to the changes in this PR.

c42f commented:
I just rebased this again to clean up some recent conflicts, and to consolidate a lot of WIP commits into a small number of consistent changes.

This reminded me of an interesting side effect of this patch which I should point out — code generated by `@test` should now be able to run in the interpreter. This used not to work because `@test` emits a `finally` block, and the lowering of `finally` previously contained a `:foreigncall` to `jl_rethrow_other` which caused the interpreter to be bypassed.

c42f committed:
Fix some GC rooting problems, add GC-SA annotations

Add missing GC static analyzer annotations to exception stack functions
and fix two problems found by the GC static analyser:
* An early return with missing GC_POP
* Missing root for the exception across a safe point in throw_internal

c42f committed:
Use JL_BT_INTERP_FRAME for interpreter frame marker in bt_data

c42f commented:
I'm starting to think the test failures here are an existing latent bug brought out by the changes in this PR. The segfault may be due to whatever is causing #29588.

The failures also seem very specific to the CI setup, I have only been able to reproduce inside a VM which is set up similarly to the Travis ubuntu image with gcc-5.5, building with `FORCE_ASSERTIONS=1`. It's definitely a reproducible bug, but a fairly tricky one.

c42f committed:
Workaround gcc-5 miscompilation with asserts and double return functions

gcc-5 appears to miscompile jl_type_infer under special circumstances,
leading to segfaults but gcc-7 and clang do fine on this.  The "special
circumstances" are that assertions are enabled when preprocessing
jl_svecref for use within jl_type_infer specifically.

There are various ways to remove the segfaults and force a correct
compilation:

* The given rearrangement

* Replacing jl_svecref in the last part of jl_type_infer with an
  equivalent function, but without the assertions.

* Removing the JL_TRY / JL_CATCH further up.

It's suspicious that the error only occurs when both the double-return
__sigsetjmp() and the no-return __assert_fail() are present inside the
function in sufficient quantities. Perhaps the compiler is failing to
analyze the interaction between these.

vchuravy commented:
@c42f Thank you for working on this! I just lost way to much time on trying to fix bugs in code that losses it stack-traces due to tasks!

c42f commented:
@vchuravy thanks, maybe you can have a look over the code? It's basically waiting on review at this stage.

@vtjnash I think the recurring CI failures I was seeing in this PR on linux were a gcc-5.5 compiler bug (so as you expected, not related to #29588). The trivial rearrangement in 250571c11628 appears to work around the problem.

vchuravy reviewed (commented):
I have been using this for the last day running some fairly complicated code with this <3

c42f commented on base/compiler/ssair/ir.jl:
```diff
@@ -318,7 +318,7 @@ function is_relevant_expr(e::Expr)
                       :foreigncall, :isdefined, :copyast,
                       :undefcheck, :throw_undef_if_not,
-                      :cfunction, :method,
+                      :cfunction, :method, :pop_exc,
```
I'm happy to change the symbol used in the AST.

For the C code I'm not so sure — names like `jl_exception_stack_t` and `jl_exception_stack_bt_size` get a bit tiresome after a while. When spelling it out is too verbose it can be useful to make a new noun, so I could compress all `exc_stack` into `excstack` in the C code and just be consistent in the use of that. I tried to avoid using `ex` as an abbreviation because that commonly stands for "expression".

c42f replied:
@vtjnash would you also prefer all the C code to also use `exception_stack` (at least for types and function/macro names)?

Obviously after massaging this PR for a few months I've gotten rather used to my "new noun" `exc_stack` and I think it's a good tradeoff (though possibly better a single word `excstack`). But renaming is easy and I can use `exception_stack` everywhere if there's a consensus.

c42f commented:
> I have been using this for the last day running some fairly complicated code with this

Excellent!

vtjnash commented on test/exceptions.jl:
```diff
@@ -0,0 +1,210 @@
+            @test length(catch_stack()) == 1
+            return
+        end
+    end
```
Is the following form also handled:
```julia
         try
             return
         catch
         finally
         end
```

c42f replied:
Should be handled currently, but I'll add a couple of extra test cases for the future.

To explain - currently these three block
```julia
try
    A
catch
    B
finally
    C
end
```

are lowered into the nested form

```julia
try
    try
        A
    catch
        B
    end
finally
    C
end
```
very early on which is nice and simple but also rather inefficient because we could get away with a single eh frame here. Something for future work perhaps ;-)

c42f replied:
Or if you're asking about the early-return-from-try form — if you're in the try block there's no exception popping required. I'll add a few tests for that too.

vtjnash reviewed (commented):
This looks very exceptional! Outstanding!

I agree with Valentin that a longer name would be preferable for `:pop_exception`. Now that the main PR is done, I don't expect people to need to type it very often.

I'm not much surprised that you ran into a JL_TRY/JL_CATCH miscompile. The way we lower try/catch makes it easy to write invalid code (per the compiler spec), and compilers are pretty unreliable at handling even the standards-permitted usages of it.

c42f commented on src/julia_internal.h:
```diff
@@ -658,6 +659,34 @@ JL_DLLEXPORT int jl_is_interpreter_frame(uintptr_t ip);
+    // Pack all stack entries into a growable buffer to amortize allocation
+    // across repeated exception handling.
+    // uintptr_t data[]; // Access with jl_excstk_raw
+#define jl_excstk_raw(stack) ((uintptr_t*)((char*)(stack) + sizeof(jl_exc_stack_t)))
```
Nice :+1:

The C trick of simply using `uintptr_t data[]` in the struct would have been neater but the C++ compiler had a fit.

c42f commented on src/julia_internal.h:
```diff
@@ -658,6 +659,34 @@ JL_DLLEXPORT int jl_is_interpreter_frame(uintptr_t ip);
+{
+    return (jl_value_t*)jl_excstk_raw(stack)[(itr)-1];
+}
+#define jl_exc_stack_bt_size(stack, itr)   ((size_t)jl_excstk_raw(stack)[(itr)-2])
```
Always a good idea.

c42f commented:
Thanks for the reviews guys, much appreciated. @vtjnash those are all good points, I'll do those cleanups as soon as I get the chance. I'll also rename the expression head `:pop_exc` to `:pop_exception`.

As mentioned I'm happy to do some renaming in the C code if there's a clear consensus. I'll play around with the naming after the other cleanups.

c42f committed:
Exception stack runtime cleanup

* Use functions rather than macros for exception stack access
* Use standard type check macro
* More clearly document gc rooting of `jl_current_exception()`

c42f committed:
Rename Expr(:pop_exc) to Expr(:pop_exception)

Better be explicit here for something that's not commonly encountered.

c42f committed:
More comprehensive exception stack tests

* Test try-catch-finally form
* Add a few tests to do with exiting try blocks.
* More explanatory comments

c42f commented:
Review comments addressed — see the three commits above.

c42f committed:
Fix gc static analyzer annotations

c42f committed:
Automated renaming exc_stack -> excstack

cd src
ag -l exc_stack  | xargs sed -i -e 's/exc_stack/excstack/g'
ag -l excstk_raw | xargs sed -i -e 's/excstk_raw/excstack_raw/g'
sed -i -e 's/jlcurrent_exception_func/jl_current_exception_func/g' codegen.cpp

cd ../base
ag -l exc_stack  | xargs sed -i -e 's/exc_stack/excstack/g'

c42f commented:
I played with naming a bit.  On balance I preferred `excstack` as a new noun than spelling out `exception_stack` everywhere in the C code, so I went with that. At 161 occurrences in the C code it seems reasonably justified. Can still be changed if necessary.

c42f commented:
@JeffBezanson I think this is ready for you to look at if you can make the time. I think I'm done at this end for now.

vtjnash reviewed (approved):
I like the renamings.

KristofferC commented on base/error.jl:
```diff
@@ -91,6 +91,30 @@ function catch_backtrace()
+arbitrary task. This is useful for inspecting tasks which have failed due to
+uncaught exceptions.
+"""
+function catch_stack(task=current_task(); include_bt=true)
```
Needs to be added somewhere to the docs?

c42f replied:
Excellent point, done (+NEWS).

c42f committed:
Documentation + NEWS for exception stacks

c42f commented:
Having written some documentation, and randomly coming across https://blog.xkcd.com/2015/09/22/a-thing-explainer-word-checker/ again, I give you the docs, simplewriter version:

## Problem rows and `catch_stack`

While fixing a problem, more problems can happen. It can be good to look at all these problems together to find out why. Julia helps you by putting each problem on the end of a row as it happens. When the plan leaves a `catch` without making more problems happen, the problems from the `try` have been fixed and they are removed from the row.

The row of problems which are happening right now can be seen by playing the `catch_stack` game.

c42f committed:
Merge remote-tracking branch 'origin/master' into cjf/exception-stack

vchuravy commented:
This looks good, one question is in which order we should merge this and #22631.
PARTR is very close to being finished. I just checked and there are indeed merge conflicts. I personally would say to hold this PR until after the PARTR merge.
@JeffBezanson and @vtjnash do you have an opinion on that?

c42f commented:
Yes all the task rearrangements in PARTR were bound to create some conflicts with this. I went through and merged the two branches locally. It's mostly cosmetic, but there's a couple of places where care are needed, particularly where large chunks of code from task.c have been copied into partr.c and the changes need to be repeated.

c42f commented:
Well, PARTR might be nearly ready but it strikes me as much more difficult and intrusive than the changes here and with CI still failing a bit randomly it's not quite clear when it will be really finished. So I tentatively suggest we just merge this rather than waiting as it seems to be stable now. In the event we merge this first, I'm happy to merge master into the PARTR branch and repair the few exception stack related things over there.

vtjnash commented:
Agreed. Let's merge this.

vchuravy committed:
Merge branch 'master' into cjf/exception-stack

vtjnash commented:
Currently this isn't being used much, right? So once we merge this, we should open an issue to port over usages of it (such as many uses of `jl_current_exception` and other places we manually call `Base.showerror`), and port them over to exposing all of the information in TLS (instead of passing it directly via `the_exception, catch_stacktrace()`).

c42f commented:
> Currently this isn't being used much, right

Yes the stack is used underneath but isn't generally exposed to the user unless they called `catch_stack` themselves. I've got some hacky WIP for exposing it but I'm not sure about what's best yet. We should definitely open an issue.

c42f commented:
@ararslan The freebsd buildbot (normally most reliable and awesome of all build bots!) is having a problem today and seems to reproducibly fail to download the pcre depedency. Restarting multiple times has not helped.

ararslan commented:
Should be fixed now. The currently running build looks poised to complete successfully.

c42f commented:
@ararslan Thanks so much!

Pending CI passing once more and the JuliaLang/julia merge lockout being removed, I will go ahead and merge this. Hopefully later today.

iamed2 commented:
Congrats on this!! I'm excited to use these when I can!
